Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+  {
+    sendDigit(o2::mch::Digit(deId, padId, adc, time));
+  }
+
+  void mergeDigits(int feeId) {}
+};
+
+//_________________________________________________________________
+class Merger : public MergerBase
+{
+  int32_t feeId = {-1};
+  FeeIdMerger mergers[MCH_MERGER_FEEID_MAX + 1];
+
+ public:
+  Merger() = default;
+  ~Merger() = default;
+
+  void setOrbit(int feeId, uint32_t orbit, bool stop);
+
+  void setDigitHandler(std::function<void(const Digit&)> h);
+
+  void addDigit(int feeId, int solarId, int dsAddr, int chAddr,
+                int deId, int padId, int adc, Digit::Time time, Digit::Time stopTime);
+
+  void mergeDigits(int feeId);
+};
+
+void FeeIdMerger::setOrbit(uint32_t orbit, bool stop)
+{
+  // send digits of previous orbit if either the stop RDH is received
+  // or a new orbit is started
+  if ((orbit == buffers[currentBufId].orbit) && (!stop)) {
+    return;
+  }
+
+  int nSent = 0;
+  for (auto& d : buffers[previousBufId].digits) {
+    if (!d.merged && (d.digit.getPadID() >= 0)) {
+      uint16_t size = (d.stopTime.getBXTime() - d.digit.getTime().getBXTime()) / 4 + 1;
+      sendDigit(d.digit);
+      nSent += 1;
+    }
+  }
+  if (mPrint) {
+    std::cout << "[FeeIdMerger] sent " << nSent << " digits for orbit " << buffers[previousBufId].orbit << "  current orbit is " << orbit << std::endl;
+  }
+
+  currentBufId = 1 - currentBufId;
+  previousBufId = 1 - previousBufId;
+  buffers[currentBufId].digits.clear();
+  buffers[currentBufId].orbit = orbit;
+}
+
+void FeeIdMerger::mergeDigits()
+{
+  auto checkDigits = [](MergerDigit& d1, MergerDigit& d2) -> bool {
+    // skip digits that are already merged
+    if (d1.merged || d2.merged)
+      return false;
+
+    // skip all digits not matching the detId/padId
+    if (d1.solarId != d2.solarId)
+      return false;
+    if (d1.dsAddr != d2.dsAddr)
+      return false;
+    if (d1.chAddr != d2.chAddr)
+      return false;
+    if (d1.digit.getDetID() != d2.digit.getDetID())
+      return false;
+    if (d1.digit.getPadID() != d2.digit.getPadID())
+      return false;
+
+    // compute time difference
+    Digit::Time startTime = d1.digit.getTime();
+    uint32_t bxStart = startTime.bunchCrossing;
+    Digit::Time stopTime = d2.stopTime;
+    uint32_t bxStop = stopTime.bunchCrossing;
+    // correct for value rollover
+    if (bxStart < bxStop)
+      bxStart += 0x100000;
+
+    uint32_t stopTimeFull = bxStop + (stopTime.sampaTime << 2);
+    uint32_t startTimeFull = bxStart + (startTime.sampaTime << 2);
+    uint32_t timeDiff = startTimeFull - stopTimeFull;
+
+    if (mPrint) {
+      std::cout << "checkDigits: " << d1.digit.getDetID() << "  " << d1.digit.getPadID() << "  " << bxStop << "  " << stopTime.sampaTime
+                << "  " << bxStart << "  " << startTime.sampaTime << "  " << timeDiff << std::endl;
+    }
+
+    // skip if the time difference is not equal to 1 ADC clock cycle
+    if (timeDiff > 8)
+      return false;
+
+    // merge digits
```

Comment:
provide a += operator (or some e.g. mergeWith member function) to the MergedDigit ? to make it clearer which of the two will be modified ? 
```
d2 += d1;
d1.merged = true;
```

Reply:
Done